stores/metadata.go

```diff
@@ -999,6 +986,148 @@ func (s *SQLStore) RenameObjects(ctx context.Context, prefixOld, prefixNew strin
 	return nil
 }
 
+func (s *SQLStore) FetchPartialSlab(ctx context.Context, ec object.EncryptionKey, offset, length uint32) ([]byte, error) {
+	key, err := ec.MarshalText()
+	if err != nil {
+		return nil, err
+	}
+
+	var data []byte
+	err = s.retryTransaction(func(tx *gorm.DB) error {
+		var buffer dbBufferedSlab
+		err := tx.
+			Joins("DBSlab").
+			Take(&buffer, "DBSlab.key = ?", key).
+			Error
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return api.ErrObjectNotFound
+		} else if err != nil {
+			return err
+		}
+		f, err := os.Open(filepath.Join(s.partialSlabDir, buffer.Filename))
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		data = make([]byte, length)
+		_, err = f.ReadAt(data, int64(offset))
+		return err
+	})
+	return data, err
+}
+
+func (s *SQLStore) AddPartialSlab(ctx context.Context, data []byte, minShards, totalShards uint8, contractSet string) (slabs []object.PartialSlab, err error) {
+	// Sanity check input.
+	slabSize := bufferedSlabSize(minShards)
+	if minShards == 0 || totalShards == 0 || minShards > totalShards {
+		return nil, fmt.Errorf("invalid shard configuration: minShards=%v, totalShards=%v", minShards, totalShards)
+	} else if contractSet == "" {
+		return nil, fmt.Errorf("contract set must not be empty")
+	} else if len(data) > slabSize {
+		return nil, fmt.Errorf("data size %v exceeds size of a slab %v", len(data), slabSize)
+	}
+
+	err = s.retryTransaction(func(tx *gorm.DB) error {
+		// Fetch contract set.
+		var cs dbContractSet
+		if err := tx.Take(&cs, "name = ?", contractSet).Error; err != nil {
+			return err
+		}
+
+		// Find a buffer that is not yet marked as complete.
+		// NOTE: We don't need to fetch the data of the buffer since we are just
+		// going to append to it. Instead we omit the data and select the length
+		// of the data.
+		var buffer struct {
+			ID       uint
+			Filename string
+			SlabID   uint
+			SlabKey  []byte
+			Size     int64
+		}
+		err := tx.
+			Table("buffered_slabs").
+			Select("buffered_slabs.id AS ID, buffered_slabs.size AS Size, buffered_slabs.filename AS Filename, sla.id AS SlabID, sla.key AS SlabKey").
+			Joins("INNER JOIN slabs sla ON buffered_slabs.id = sla.db_buffered_slab_id AND sla.min_shards = ? AND sla.total_shards = ?", minShards, totalShards).
+			Joins("INNER JOIN contract_sets cs ON sla.db_contract_set_id = cs.id AND cs.name = ?", contractSet).
+			Where("buffered_slabs.complete = ?",
+				false).
+			Take(&buffer).
+			Error
+		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+			return err
+		} else if errors.Is(err, gorm.ErrRecordNotFound) {
+			// No buffer found, create a new one.
+			slab, err := createSlabBuffer(tx, cs.ID, data, minShards, totalShards, s.partialSlabDir)
+			slabs = []object.PartialSlab{slab}
+			return err
+		}
+
+		// We have a buffer. Sanity check it.
+		if buffer.Size >= int64(slabSize) {
+			return fmt.Errorf("incomplete buffer with ID %v has no space left, this should never happen", buffer.ID)
```

Comment:
`nit` wouldn't bubble up a "this should never happen" here as it turns into a user-facing error I believe. It's a classic log line in a GH issue when they come showing us "this should never happen" errors and it's embarrassing